README.md
# [Kern Valley Camping](https://camping.kernvalley.us)
Map of camping sites in and around the Kern River Valley

<!-- [![Dependabot Status](https://api.dependabot.com/badges/status?host=github&repo=kernvalley/camping.kernvalley.us)](https://dependabot.com) -->
[![Node CI](https://github.com/kernvalley/camping.kernvalley.us/workflows/Node%20CI/badge.svg)](https://github.com/kernvalley/camping.kernvalley.us/actions)
[![Super Linter](https://github.com/kernvalley/camping.kernvalley.us/workflows/Lint%20Code%20Base/badge.svg)](https://github.com/kernvalley/camping.kernvalley.us/actions?query=workflow%3A%22Lint+Code+Base%22)
[![Netlify Status](https://api.netlify.com/api/v1/badges/252539d1-4596-41e9-9d63-97a964822b25/deploy-status)](https://app.netlify.com/sites/infallible-galileo-ac41ee/deploys)

[![GitHub license](https://img.shields.io/github/license/kernvalley/camping.kernvalley.us.svg)](https://github.com/kernvalley/camping.kernvalley.us/blob/master/LICENSE)
[![GitHub issues](https://img.shields.io/github/issues/kernvalley/camping.kernvalley.us.svg)](https://github.com/kernvalley/camping.kernvalley.us/issues)
[![GitHub pull requests](https://img.shields.io/github/issues-pr/kernvalley/camping.kernvalley.us.svg)](https://github.com/kernvalley/camping.kernvalley.us/pulls)
[![GitHub last commit](https://img.shields.io/github/last-commit/kernvalley/camping.kernvalley.us.svg)](https://github.com/kernvalley/camping.kernvalley.us/commits/master)
[![GitHub release](https://img.shields.io/github/release/kernvalley/camping.kernvalley.us.svg)](https://github.com/kernvalley/camping.kernvalley.us/releases)

![GitHub followers](https://img.shields.io/github/followers/kernvalley.svg?style=social)
![GitHub forks](https://img.shields.io/github/forks/kernvalley/camping.kernvalley.us.svg?style=social)
![GitHub stars](https://img.shields.io/github/stars/kernvalley/camping.kernvalley.us.svg?style=social)
[![Twitter Follow](https://img.shields.io/twitter/follow/kern_valley.svg?style=social)](https://twitter.com/krn_valley)

[![Donate using Liberapay](https://img.shields.io/liberapay/receives/shgysk8zer0.svg?logo=liberapay)](https://liberapay.com/shgysk8zer0/donate "Donate using Liberapay")
![Keybase BTC](https://img.shields.io/keybase/btc/shgysk8zer0.svg)
![Keybase PGP](https://img.shields.io/keybase/pgp/shgysk8zer0.svg)
- - -

- [Code of Conduct](./.github/CODE_OF_CONDUCT.md)
- [Contributing](./.github/CONTRIBUTING.md)
<!-- - [Security Policy](./.github/SECURITY.md) -->


CHANGELOG.md
<!-- markdownlint-disable -->
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v2.0.1] - 2023-07-05

### Changed
- Update dependencies and config

## [v2.0.0] - 2023-05-12

### Added
- Implement import maps
- Implement trusted types
- Button for fullscreen
- Button to show user's current location

### Changed
- Re-write to use components in `_includes/common/` and `_includes/leaflet/` via submodules
- Use main site privacy page and add redirect

### Removed
- Git copy of `icons.svg` (rely on `svg-sprite-generator` on build)

## [v1.1.1] - 2021-01-17

### Changed
- Update components and places data

## [v1.1.0] - 2021-01-11

### Added
- Places data as submodule
- New and updated components in `_includes/` for map markers
- Weather from OpenWeatherMap via `<weather-current>`
- Also show markers for towns, gas stations, stores, etc.
- Add fullscreen button

### Changed
- Use data from places instead of local copy of data
- Replace contact page with redirect to `https://contact.kernvalley.us`
- Update nav layout

### Removed
- Remove unused icons from `icons.svg`
- Clean-up SW cache list
- `<html-notification>` notifying of site being under construction

### Fixed
- Footer layout in Chrome (do not use multi-column layout)
- Fix incorrect viewport / nav placement on mobile

## [v1.0.6] - 2020-12-13

### Added
- Add `<button is="app-list">`
- Control theme via `cookieStore`

### Changed
- Move `<button is="pwa-install">` to `<nav>`

### Changed
- Update to Leaflet [1.7.1](https://leafletjs.com/2020/09/04/leaflet-1.7.1.html)

## [v1.0.4] - 2020-09-06

### Added
- "Under Construction" notification
- Preloading of assets
- Different sized background images
- GA tracking of external link clicks

### Changed
- Various dependency & workflow updates
- Import updates to some components

### Fixed
- Add missing search icon
- Disable bash script linting due to issue with `.rvmrc`

### Removed
- Duplicate form styles (use CDN CSS `.form-group` rules)

## [v 1.0.3] - 2020-07-18

### Changed
- Use `history.state` for managing map navigation
- Specify indent style and width in [`.editorconfig`](https://editorconfig.org/)
- Update various config files
- Update icons to be compatible with maskable icons
- Disable `'unsafe-inline'` for styles in CSP

## [v1.0.2] - 2020-07-15

### Updated
- Misc. changes for components to use external stylesheets

## [v1.0.1] - 2020-06-30

### Fixed
- Invalid call to `HTMLLeafletMarkerElement.close()` - use `open = false`

## [v1.0.0] - 2020-06-30

### Added
- Google Analytics
- Data file for campgrounds
- Privacy Policy
- Contact Page
- App icons & TWA asset links
- Search handler
- Share buttons
- Submodules now watched by dependabot
- Update URL hash on popup close [#5](https://github.com/kernvalley/camping.kernvalley.us/issues/5)

### Fixed
- Correctly set `<meta property="og:image">`

### Changed
- Theme colors
- Misc. metadata
<!-- markdownlint-restore -->


sw-config.js
---
layout: null
---
'use strict';
/*eslint no-unused-vars: 0*/
const config = {
	version: '{{ site.data.app.version | default: site.version }}',
	fresh: [
		'/',
		'https://apps.kernvalley.us/apps.json',
		'https://cdn.kernvalley.us/img/markers.svg',
		'/manifest.json',
	].map(path => new URL(path, location.origin).href),
	stale: [
		/* Other HTML */
		'{{ site.data.importmap.imports["@shgysk8zer0/components/"] }}leaflet/map.html',
		'{{ site.data.importmap.imports["@shgysk8zer0/components/"] }}github/user.html',
		'{{ site.data.importmap.imports["@shgysk8zer0/components/"] }}weather/current.html',
		'{{ site.data.importmap.imports["@shgysk8zer0/components/"] }}toast-message.html',
		'{{ site.data.importmap.imports["@shgysk8zer0/components/"] }}install/prompt.html',

		/* JS, `customElements`, etc. */
		'/js/index.min.js',
		'{{ site.data.importmap.imports["@shgysk8zer0/polyfills"] }}',
		'{{ site.data.importmap.imports["@shgysk8zer0/components/"] }}leaflet/bundle.min.js',

		/* CSS */
		'/css/index.min.css',
		'{{ site.data.importmap.imports["leaflet/"] }}dist/leaflet.css',
		'{{ site.data.importmap.imports["@shgysk8zer0/components/"] }}leaflet/map.css',
		'{{ site.data.importmap.imports["@shgysk8zer0/components/"] }}github/user.css',
		'{{ site.data.importmap.imports["@shgysk8zer0/components/"] }}weather/current.css',
		'{{ site.data.importmap.imports["@shgysk8zer0/components/"] }}toast-message.css',
		'{{ site.data.importmap.imports["@shgysk8zer0/components/"] }}install/prompt.css',

		/* Images & Icons */
		'/img/icons.svg',
		'/img/apple-touch-icon.png',
		'/img/icon-512.png',
		'/img/icon-192.png',
		'/img/icon-32.png',
		'/img/favicon.svg',
		'https://cdn.kernvalley.us/img/adwaita-icons/actions/mark-location.svg',
		'https://cdn.kernvalley.us/img/logos/play-badge.svg',
		'https://cdn.kernvalley.us/img/keep-kern-clean.svg',
		'https://cdn.kernvalley.us/img/logos/instagram.svg',

		/* Fonts */
		'https://cdn.kernvalley.us/fonts/roboto.woff2',
		/* Other */
	].map(path => new URL(path, location.origin).href),
	allowed: [
		// /https:\/\/maps\.wikimedia\.org\/osm-intl\/*/,
		// /https:\/\/[A-z]\.tile\.openstreetmap\.org\/*/,
		'https://server.arcgisonline.com/ArcGIS/rest/services/',
		'https://i.imgur.com/',
		/https:\/\/*\.githubusercontent\.com\/u\/*/,
		/\.(jpg|png|gif|svg|webp)$/,
	],
	allowedFresh: [
		'https://api.openweathermap.org/data/',
		'https://api.github.com/users/',
		/\.(html|css|js|json)$/,
	]
};


js/handlers.js
import { $, createCustomElement } from 'https://cdn.kernvalley.us/js/std-js/functions.js';
import { site } from './consts.js';

export async function stateHandler({ state }) {
	const { uuid = null, longitude = NaN, latitude = NaN, title = null, body = null } = state || {};

	if (typeof uuid === 'string') {
		const marker = document.getElementById(uuid);

		if (marker instanceof HTMLElement && marker.tagName === 'LEAFLET-MARKER') {
			const map = marker.closest('leaflet-map');
			await map.ready;
			document.title = `${title} | ${site.title}`;
			map.center = marker;
			marker.open = true;
		}
	} else if (! (Number.isNaN(longitude) || Number.isNaN(latitude))) {
		const marker = await createCustomElement('leaflet-marker');
		const map = document.querySelector('leaflet-map');
		const icon = new Image(32, 32);
		await map.ready;

		icon.decoding = 'async';
		icon.loading = 'lazy';
		icon.alt = '';
		icon.referrerPolicy = 'no-referrer';
		icon.crossOrigin = 'anonymous';
		icon.src = new URL('./img/adwaita-icons/actions/mark-location.svg', site.iconBaseUri).href;
		icon.slot = 'icon';

		marker.longitude = longitude;
		marker.latitude = latitude;
		marker.slot = 'markers';
		marker.title = title;

		if (body instanceof HTMLElement) {
			body.slot = 'popup';
			body.append(document.createElement('hr'), await getShareButton({text: title}));
			if ('part' in body) {
				body.part.add('popup');
			}
			marker.append(icon, body);
			map.center = marker;
			marker.open = true;
		} else if (typeof body === 'string') {
			const popup = document.createElement('div');
			const h4 = document.createElement('h4');
			const content = document.createElement('div');

			h4.textContent = title;
			content.textContent = body;
			popup.slot = 'popup';
			popup.append(h4, content, document.createElement('hr'), await getShareButton({text: title}));

			if ('part' in popup) {
				popup.part.add('popup');
			}
			marker.append(icon, popup);
			map.center = marker;
			map.append(marker);
			marker.open = true;
		} else {
			marker.append(icon);
			map.center = marker;
			map.append(marker);
			marker.open = true;
		}
	} else if (state === null) {
		document.title = `Home | ${site.title}`;
	}
}

export function hashChange({oldURL = ''} = {}) {
	if (oldURL.includes('#')) {
		const [,hash = ''] = oldURL.split('#', 2);

		if (hash !== '') {
			const target = document.getElementById(hash);

			if (target instanceof HTMLElement && target.tagName === 'LEAFLET-MARKER') {
				target.open = false;
			}
		}
	}

	if (location.hash !== '') {
		const marker = document.getElementById(location.hash.substr(1));

		if (marker instanceof HTMLElement && marker.tagName === 'LEAFLET-MARKER') {
			marker.hidden = false;
			marker.open = true;
		} else {
			$('leaflet-marker').forEach(el => el.open = false);
		}
	}
}

export function markerOpen() {
	if (this.id !== '') {
		location.href = `#${this.id}`;
	}
}

export function markerClose() {
	if (location.hash.substr(1) === this.id) {
		location.hash = '';
	}
}

export function markerSearch() {
	const value = this.value.toLowerCase();

	if (value !== '') {
		const matches = $('leaflet-marker').filter(camp => camp.title.toLowerCase().contains(value));
		matches.forEach(match => match.hidden = false);
	}
}

export async function getShareButton({text, url = location.href, textContent = 'Share'} = {}) {
	const btn = await createCustomElement('share-button');
	btn.text = text;
	btn.url = url;
	btn.textContent = textContent;
	btn.hidden = ! (navigator.share instanceof Function);
	return btn;
}


js/functions.js
export function getMarkers() {
	return Array.from(document.querySelectorAll('leaflet-marker[itemtype="https://schema.org/Campground"], leaflet-marker[itemtype="https://schema.org/City"]'));
}

export function getMarkerTitles() {
	return getMarkers().map(marker => marker.title);
}

export function registerMapSearch(form = document.forms.search) {
	if (form instanceof HTMLFormElement) {
		form.hidden = false;
		form.addEventListener('submit', async event => {
			event.preventDefault();
			const data = new FormData(event.target);
			const query = data.get('query').toLowerCase();

			if (query !== '') {
				const markers = getMarkers();
				markers.forEach(marker => marker.hidden = ! marker.title.toLowerCase().includes(query));

				const shown = markers.filter(marker => marker.hidden === false);
				if (shown.length === 1) {
					shown[0].open = true;
				}
			}
		});

		form.addEventListener('reset', () => {
			getMarkers().forEach(el => el.hidden = el.classList.contains('closed'));
		});

		const datalist = form.querySelector('#campgrounds-list');

		if (datalist instanceof Element) {
			const opts = getMarkerTitles().map(name => {
				const opt = document.createElement('option');
				opt.value = name;
				return opt;
			});

			datalist.append(...opts);
		}
	}
}


js/index.js
import '@shgysk8zer0/kazoo/theme-cookie.js';
import { createPolicy } from '@shgysk8zer0/kazoo/trust.js';
import { getGooglePolicy, getDefaultPolicy } from '@shgysk8zer0/kazoo/trust-policies.js';
import { debounce } from '@shgysk8zer0/kazoo/events.js';
import { ready, on, css, toggleClass, each, map } from '@shgysk8zer0/kazoo/dom.js';
import { getCustomElement } from '@shgysk8zer0/kazoo/custom-elements.js';
import { init } from '@shgysk8zer0/kazoo/data-handlers.js';
import { loadImage } from '@shgysk8zer0/kazoo/loader.js';
import { importGa, externalHandler, telHandler, mailtoHandler } from '@shgysk8zer0/kazoo/google-analytics.js';
import { SECONDS } from '@shgysk8zer0/kazoo/date-consts.js';
import { site, GA } from './consts.js';

getDefaultPolicy();

toggleClass([document.documentElement], {
	'no-dialog': document.createElement('dialog') instanceof HTMLUnknownElement,
	'no-details': document.createElement('details') instanceof HTMLUnknownElement,
	'js': true,
	'no-js': false,
});


if (! CSS.supports('height', '100dvh')) {
	css([document.documentElement], { '--viewport-height': `${window.innerHeight}px` });
	on([window], ['resize'], debounce(() =>
		css([document.documentElement], { '--viewport-height': `${window.innerHeight}px` })
	), { passive: true });
}

getCustomElement('install-prompt').then(HTMLInstallPromptElement => {
	on('#install-btn', ['click'], () => new HTMLInstallPromptElement().show())
		.forEach(el => el.hidden = false);
});

if (typeof GA === 'string' && GA.length !== 0) {
	const policy = getGooglePolicy();
	scheduler.postTask(async() => {
		const { ga, hasGa } = await importGa(GA, {}, { policy });

		if (hasGa()) {
			ga('create', GA, 'auto');
			ga('set', 'transport', 'beacon');
			ga('send', 'pageview');

			on('a[rel~="external"]', 'click', externalHandler, { passive: true, capture: true });
			on('a[href^="tel:"]', 'click', telHandler, { passive: true, capture: true });
			on('a[href^="mailto:"]', 'click', mailtoHandler, { passive: true, capture: true });
		}
	}, { priority: 'background' });
} else {
	createPolicy('goog#html', {});
	createPolicy('goog#script-url', {});
	getGooglePolicy();
}

ready().then(() => {
	init();

	if (location.pathname === '/') {
		Promise.all([
			getCustomElement('leaflet-marker'),
			customElements.whenDefined('leaflet-map'),
		]).then(([LeafletMarker]) => {
			on('#query', ['input'], debounce(({ target: { value: query }}) => {
				if (query.length === 0) {
					each('leaflet-marker[title][hidden]', el => el.hidden = false);
				} else {
					const queryStr = query.toLowerCase();
					each('leaflet-marker[title]', marker => {
						marker.hidden = ! marker.title.toLowerCase().includes(queryStr);
					});
				}

				const openMarkers = document.querySelectorAll('leaflet-marker:not([hidden])');

				if (openMarkers.length === 1) {
					const marker = openMarkers.item(0);
					const map = marker.closest('leaflet-map');
					marker.open = true;
					map.flyTo({
						latitude: marker.latitude,
						longitude: marker.longitude,
						zoom: 16,
					});
				}
			}));

			on([document.forms.search], ['reset'], () => {
				each('leaflet-marker[title][hidden]', el => el.hidden = false);
			});

			on([document.forms.search], ['submit'], event => event.preventDefault());

			document.forms.search.hidden = false;

			on('#find-btn', ['click'], async () => {
				const map = document.querySelector('leaflet-map');
				const ShareButton = await getCustomElement('share-button');
				const { coords: { latitude, longitude }} = await map.locate({
					enableHighAccuracy: true,
					maxAge: 15 * SECONDS,
					maxZoom: 16,
				});
				const share = new ShareButton();
				const popup = document.createElement('div');
				const h3 = document.createElement('h3');
				const pos = document.createElement('div');
				const url = new URL(document.baseURI);
				const icon = await loadImage('@shgysk8zer0/img/markers.svg#map-marker', {
					height: 30,
					width: 30,
					slot: 'icon',
					loading: 'lazy',
				});
				url.hash = `#${latitude},${longitude}`;
				h3.textContent = 'Current Location';
				pos.textContent = `${latitude}, ${longitude}`;
				share.textContent = 'Share Location';
				share.url = url.href;
				share.text = 'See my current location';

				popup.append(h3, pos, share);

				const marker = new LeafletMarker({ latitude, longitude, icon, popup });
				marker.addEventListener('close', ({ target }) => target.remove());
				document.title = `Current Location: ${site.title}`;
				marker.open = true;
				map.append(marker);
			});

			const list = document.getElementById('campgrounds-list');

			new Set(map('leaflet-marker[title]', ({ title }) => title)).forEach(value => {
				const option = document.createElement('option');
				option.value = value;
				list.append(option);
			});
		});
	}
});
